### PR TITLE
validataion for voice response question

### PR DESCRIPTION
**Pull request by huangjosh**

### CODE REVIEW

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx [hunk 1]
```diff
@@ -54,10 +64,42 @@ class VisibleSubmissionEditIndex extends Component {
...
+    let readyToSubmit = true;
+    Object.values(form.syncErrors || {}).forEach((answerError) => {
+      const noError = Object.keys(answerError).length === 0;
+      if (!noError) {
+        readyToSubmit = false;
+      }
+    });
+
+    if (!readyToSubmit && showErrorNotification) {
+      setNotification(translations.submitError)(dispatch);
```
Comment (allenwq) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx]: Would be better if you can take the notification logic out and keep the method single responsibility. Now this method returns a value + set the notification.

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx [hunk 2]
```diff
@@ -54,10 +64,42 @@ class VisibleSubmissionEditIndex extends Component {
...
+      const answerId = answer.id;
+      Object.keys(answer).forEach((key) => {
+        dispatch(touch(formNames.SUBMISSION, `${answerId}.${key}`));
+      });
+    });
+
+    let readyToSubmit = true;
+    Object.values(form.syncErrors || {}).forEach((answerError) => {
+      const noError = Object.keys(answerError).length === 0;
+      if (!noError) {
```
Comment (allenwq) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx]:
```
const hasError = ...
if (hasError)
````

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx [hunk 3]
```diff
@@ -54,10 +64,42 @@ class VisibleSubmissionEditIndex extends Component {
...
   }
 
+  /**
+   * Return true if it is ready to submit
+   */
+  validateSubmit = (showErrorNotification = true) => {
+    const { dispatch, form } = this.props;
+    const answers = Object.values(form.values);
+
+    // touch all the values
```
Comment (allenwq) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx]: Can add more comments on why need to touch all answers?

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    defaultMessage: 'Please record your voice',
+  },
+});
+
+const validateDefaultQuestion = () => undefined;
+
+const validateVoiceResponse = (values) => {
+  const { file: fileValue } = values;
+  const { file, url } = fileValue;
+  if (url || Object.prototype.toString.call(file) === '[object File]') {
```
Comment (allenwq) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js]: Any better ways to check it is a file?

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import { questionTypes } from '../../constants';
+
+const translations = defineMessages({
+  pleaseRecordYourVoice: {
+    id: 'course.assessment.submission.pages.SubmissionEditIndex.submissionFormValidate.pleaseRecordYourVoice',
+    defaultMessage: 'Please record your voice',
+  },
+});
+
+const validateDefaultQuestion = () => undefined;
```
Comment (allenwq) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js]: Prefer to preserve `undefined` to those variables that are really undefined. perhaps we just return `true` by default?

**Comment by huangjosh** [comment]
@allenwq Fixed and added simple test case.

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js (hunk 4 above)
Comment (allenwq) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js]: @huangjosh any updates on this?

Comment (huangjosh) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/submissionFormValidate.js]: @allenwq I think ```instanceof``` would be a better choice. Fixed, and updated. the PR.

**Comment by allenwq** [comment]
Thanks @huangjosh !

**Comment by huangjosh** [comment]
@allenwq No problem. :)

**Pull request merged**